KAFKA-13999: Add ProducerCount metrics (KIP-847)

anatasiavela opened the pull request:
This is the PR for the implementation of KIP-847: https://cwiki.apache.org/confluence/display/KAFKA/KIP-847%3A+Add+ProducerIdCount+metrics
Add ProducerIdCount metric at the broker level:
```
kafka.server:type=ReplicaManager,name=ProducerIdCount
```

Added unit tests below to ensure the metric reported the count correctly.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

artemlivshits committed:
KAFKA-13999: Add ProducerCount metrics (KIP-847)

jolshan commented:
Thanks @anatasiavela and @artemlivshits -- I'll take a look at the build when it completes, but this looks pretty good so far.

divijvaidya commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -576,6 +576,13 @@ class UnifiedLog(@volatile var logStartOffset: Long,
   }, period = producerIdExpirationCheckIntervalMs, delay = producerIdExpirationCheckIntervalMs, unit = TimeUnit.MILLISECONDS)
 
+  // Visible for testing
+  def removeExpiredProducers(currentTimeMs: Long): Unit = {
```
could we use this method at producerExpireCheck() as well?

anatasiavela replied:
yup, good catch. Reused that method.

divijvaidya commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -685,6 +692,7 @@ class ProducerStateManager(
 
     if (logEndOffset != mapEndOffset) {
       producers.clear()
+      _producerIdCount = 0
```
Please move this together in a function. This coupling between the metric value `producers` is bug prone where someone may accidentally update one without updating the other. One way to avoid such bugs in future to always update the producer map using methods where count would be updated too (ideally only those functions should mutate the number of producers). 

Similar comment for adding to the `producers` as well.

anatasiavela replied:
good point. I've coupled the two and made the change as you suggested.

divijvaidya reviewed (commented):
Please consider documenting this metric in docs/ops.html (with a description of the metric)

anatasiavela committed:
couple producers map and metric value

anatasiavela commented:
Thank you for the review @divijvaidya! I've addressed your comments and made changes accordingly, so please take a look whenever you get the chance. Thanks!

jolshan commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -684,7 +700,7 @@ class ProducerStateManager(
     }
 
     if (logEndOffset != mapEndOffset) {
-      producers.clear()
```
Maybe we can have one more helper for clear so we don't need to remove one by one?

anatasiavela replied:
I considered another helper, but `clear()` iterates all the keys anyways so I thought it would be cleaner this way
```
def clear(): Unit = { keysIterator foreach -= }
```

jolshan replied:
Oh I see. Thanks for teaching me something :)

anatasiavela committed:
update docs/ops.html

divijvaidya reviewed (approved):
Looks good to me!

clolov commented:
Is there a reason we don't change the HashMap to a Concurrent*Map and just call `size()` on it? This way it becomes very clear that we expect some sort of a concurrent operation to happen on the map and we would like it to be ordered with respect to other operations and we do not have to rely on people knowing they need to interact with the `producers` map via additional functions?

divijvaidya commented on docs/ops.html:
```diff
@@ -1604,6 +1604,11 @@ <h4 class="anchor-heading"><a id="remote_jmx" class="anchor-link"></a><a href="#
+      <tr>
+        <td>Producer Id counts</td>
+        <td>kafka.server:type=ReplicaManager,name=ProducerIdCount</td>
+        <td>Count of all producer ids for each replica on the broker.</td>
```
From the KIP, this should probable be:
> The total number of active transactional / idempotent producer ids in all partitions maintained in the broker.

jolshan replied:
I think what is written here is fine. If we want to change the wording to "in" instead of "for" that may slightly improve the meaning. 

Active here is a bit subjective, since producer IDs may be inactive but not expired yet :)

divijvaidya replied:
My only concern is that non-expert user may not understand that producer IDs are not generated for all producers. It would be nice if we can explicitly clarify here that this metrics only counts the producers that create a producer ID i.e. idempotent/transactional.

jolshan replied:
I guess if we want to be really pedantic -- the producer IDs here are for idempotent producers, its just that transactions require idempotent producers. Transactional IDs are not stored here, so the original comment could also be a bit confusing. 

If we want to say transactional and idempotent producers create producer IDs after the original sentence we can do that.

divijvaidya replied:
That would be helpful. Thank you :)

jolshan commented:
[clolov](https://github.com/clolov) I see your point about people not knowing to use those methods -- maybe we can add some documentation. 

However, I'm wondering if changing the map will affect performance for the reading and writing of the map. I don't think it makes sense to take a hit there. We will also be reading the size very frequently for the metric.

divijvaidya commented:
> Is there a reason we don't change the HashMap to a Concurrent*Map and just call `size()` on it? This way it becomes very clear that we expect some sort of a concurrent operation to happen on the map and we would like it to be ordered with respect to other operations and we do not have to rely on people knowing they need to interact with the `producers` map via additional functions?

Using ConcurrentMaps would be ideal (we won't have to acquire a lock to perform read), but the `size()` operation of a concurrent map may produce stale data (case when it's being modified concurrently). The current approach guarantees that size is strongly consistent with the data stored in the map.

jolshan commented:
@anatasiavela Could you rebase the PR?

ijuma commented:
> The current approach guarantees that size is strongly consistent with the data stored in the map.

That is not required for a metric though.

That said, I didn't look at the details of the map we are discussing here.

divijvaidya commented:
**Proposal: Using concurrent map for ProducerStateManager.producers (currently a mutable.Map)**

**Pros**
- Simplifies code (prevents future bugs by accidental update to map without updating metrics) to implement KIP-847
- Does not have any locking concerns since calculation of size() in a concurrent map does't require locks. [From Java docs for `concurrentMap`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/ConcurrentHashMap.html):
> Bear in mind that the results of aggregate status methods including size, isEmpty, and containsValue are typically useful only when a map is not undergoing concurrent updates in other threads. Otherwise the results of these methods reflect transient states that may be adequate for monitoring or estimation purposes, but not for program control.

**Cons**
- The metrics for number of producers may be inaccurate when multiple threads are modifying the map concurrently. `producers` map is modified during `truncation` (when the replica is catching up) or when a producer is expired due to timeout or on every completed transaction. That means it is modified fairly frequently. Hence, possibility of the `producerId` metric displaying inaccurate (usually off-by-one since there are at max 2 threads modifying this map) information is fairly high.

**My opinion**
In cases when we have just one (or two) producers, this metric would be highly unreliable (not just stale) as it provides an "approximation" of size(). It is not un-common to produce data from limited set of producers (with a large number of consumers) and hence, I would incline towards sticking to current approach of keeping this metric accurate.

Having said that, I don't have a strong opinion here. We can always come back and make this metric accurate if we find that inaccuracy is causing problems while troubleshooting.

ijuma commented:
Has anyone checked that we always acquire a lock when we call methods like `activeProducers` and `isEmpty`? I wonder if this class has thread safety bugs.

jolshan commented:
> Has anyone checked that we always acquire a lock when we call methods like activeProducers and isEmpty? I wonder if this class has thread safety bugs.

@ijuma I actually did take a peak when working on KIP-890 and other producer ID fun. There is a lock on every usage of active producers (except its usage in a log message). 

As for isEmpty, there are two usages (one in UnifiedLog -- to rebuild the log and one in LogLoader) and neither is protected via lock.

jolshan commented:
@divijvaidya I'm also a bit confused by this comment:
> In cases when we have just one (or two) producers, this metric would be highly unreliable (not just stale) as it provides an "approximation" of size(). It is not un-common to produce data from limited set of producers (with a large number of consumers) and hence, I would incline towards sticking to current approach of keeping this metric accurate.

We will keep the IDs of any producer used within the last day or so. If there are two producers running in a steady state, the metric will stay steady.

ijuma commented:
@jolshan From what you're saying, there are some thread safety bugs, but their impact is likely minor and hence why we haven't noticed them. From my perspective, we really need to document the concurrency model for this class more clearly and the class would be more robust if we made it thread safe.

divijvaidya commented:
@ijuma @jolshan what I wanted to convey was a situation where the approximate nature of this metrics may cause problem during troubleshooting. **But I was wrong. I read the code incorrectly at one place and wrongly assumed that this map is updated frequently**

Let me write it as a FAQ.

**What is the situation that you are worried about?**
When we have only one or two producers, the approximate nature of the metric may display the producerId to be 0 when actually it should be 1. Having one producer is a realistic production scenario since there are use cases for Kafka where folks use Kafka primarily for fan out, i.e. single producer but a large number of consumers.

**Would the metric be inconsistent only during concurrent updates to `producers` map (and stable for rest of the time)?**
Yes

**What is the likelihood that the metric is captured by the gauge at the same time when the map is undergoing concurrent update?**
Not likely (this is where I was wrong in my earlier comment). Note that producer IDs are *not* added frequently and the map is modified only during expiration or addition. Hence, the likelihood of metric display at the same time when map is being mutated is low.

**Is the metric only going to be stale (reflecting a state after some time of it's occurrence) or is it going to be inaccurate (even after some time of state change it may reflect the wrong state)?**
It can be inaccurate. This is because when modifications are being done on concurrent hash map, they are pointer manipulation and depending on the state the internal data structure the size may reflect an inaccurate value.

I don't have any further concerns with using @clolov's proposal to use ConcurrentHashMap, in fact, that is the better approach here.

Thanks all for clarifying my doubts and answering my questions. Appreciate it.

anatasiavela commented:
@divijvaidya Since this PR is strictly focusing on the metric, I think it's reasonable for that change to be done in a separate PR. what do you think?

anatasiavela committed:
add detail to metric description

divijvaidya commented:
@anatasiavela the proposal (using concurrentHashMap for producers) impacts this change. Let me try to explain why.

Currently you are using `producerIdCount` and updating it on every add/delete to the `producers` map. Alternative path was to simply call `producers.size()` whenever we want to compute the metric. But you didn't choose this path because `producers` is accessed by multiple threads at the same time and we usually guard access to it via a lock. Hence, if you wanted to call `producers.size()` you would have to acquire a lock and that is not optimal for a mere metric calculation. Hence, your approach made sense. BUT if you change the `producers` to a concurrentHashMap, you wouldn't have to acquire a lock to call `producers.size()` when computing for metrics. Without acquiring a lock, the risk is that another thread may be mutating the `producers` map at the same time. As per the discussion above in the PR, that is acceptable to us. Hence, you can simply change the `producers` to a concurrentHashMap and remove the code logic to update `producerIdCount` on every add/remove. This simplifies the code greatly.

Let me know if that makes sense.

Hangleton commented:
Thanks @anatasiavela for the PR. Apologies for any repeat, just wanted to ask if you would plan to have this merged before or after [PR-13040](https://github.com/apache/kafka/pull/13040)? Many thanks!

jolshan commented:
@Hangleton We can try to get this merged before https://github.com/apache/kafka/pull/13040 if that helps.

jolshan commented:
@ijuma Do you have any concerns with switching to ConcurrentHashMap here?

jolshan commented:
One concern is that the ConcurrentHashMap could take up more space. This metric was introduced to try to get this memory usage under control.

divijvaidya commented:
> One concern is that the ConcurrentHashMap could take up more space. This metric was introduced to try to get this memory usage under control.

The hash map in question here stores the number of idempotent producers. Which I assume is in "hundreds" in a busy cluster. As per this old benchmark https://stackoverflow.com/a/11104401 (newer JVMs may have even smaller footprint) we have an incremental usage of 8 bytes per entry, so we are talking about 800 bytes more? I think this is a reasonable tradeoff compared the out-of-the-box thread safety & code simplification that `ConcurrentHashMap` provides here. The tradeoff becomes even more appealing since the class in question, is accessed by multiple threads.

artemlivshits commented:
> Which I assume is in "hundreds" in a busy cluster

We had cases of OOM due to number of producer ids, so we definitely don't want to increase memory consumption (and should look for ways to reduce it).

dajac commented:
Would it make sense to decouple the switch to a ConcurrentHashMap from adding this metric? Given that there is no consensus, we could tackle this larger change separately if we want to. The current approach seems reasonable to me. What do you think?

ijuma commented:
Given the memory usage concerns, I think it's fine to decouple the two things. That said, we need to port this change to the Java implementation of `ProducerStateManager` as that PR has since been merged.

ijuma commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -504,7 +520,7 @@ class ProducerStateManager(
    */
   def truncateFullyAndStartAt(offset: Long): Unit = {
-    producers.clear()
+    removeProducerIds(producers.keySet)
```
It's much cheaper to clear a map than to remove all the keys. It probably makes sense to have a `clearProducerIds` method as well.

rishiraj88 replied:
Yes, nice suggestion about `clearProducerIds`. Thanks, @ijuma .

jolshan replied:
I asked Anastasia about this offline and she said the clear method also removes all the keys (using the -= operator for each one)

jolshan replied:
```{ keysIterator foreach -= }```

ijuma replied:
Not true, `java.util.HashMap`:

```java
public void clear() {
        Node<K,V>[] tab;
        modCount++;
        if ((tab = table) != null && size > 0) {
            size = 0;
            for (int i = 0; i < tab.length; ++i)
                tab[i] = null;
        }
    }
```

jolshan replied:
producers is currently defined as a mutable scala map. That's where I pulled the method from

ijuma replied:
Not in trunk. :) Generally, you should not rely on a particular inefficient implementation for this exact reason.

anatasiavela replied:
Justine originally did point this out to me but as it previously was a mutable scala map, we had kept it is. Since it's changed it makes sense to have a clearProducerIds method as well.

ijuma commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -213,6 +216,19 @@ class ProducerStateManager(
+    _producerIdCount = producers.size
+  }
+
+  def removeProducerIds(toBeRemoved: collection.Set[Long]): Unit = {
```
Why are these methods public?

anatasiavela replied:
no reason, they should be private

Hangleton commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -264,6 +265,8 @@ class ReplicaManager(val config: KafkaConfig,
     leaderPartitionsIterator.count(_.hasLateTransaction(currentTimeMs))
   }
 
+  def producerIdCount: Int = onlinePartitionsIterator.map(_.producerIdCount).sum
```
On multi-tenant clusters, having the breakdown of producer IDs by topic is important. What is the rationale to expose this metric at an aggregate level only?

jolshan replied:
Topic level metrics are a bit tricky because each broker may not host all the partitions of a give topic. However, I suppose it could give a general idea of tenant level usage. However, one concern is cardinality and the way that we would aggregate and store these topic level metrics. KIP-847 also mentions the rejected alternative of partition level metrics, but another concern there was cardinality.

Hangleton replied:
Thanks, got it. Exposing a partial aggregate per topic per broker may be misleading indeed, and we can use the new RPC `DescribeProducers` from KIP 847 can be used to get the additional level of observability.

Hangleton commented:
Thanks @jolshan for following-up. Not a problem about the merge. I was just mentioning it so that we know either PR will have to be rebased and updating accordingly.

jolshan commented:
@anatasiavela do you mind rebasing again 😅

anatasiavela committed:
fix checkstyle

jolshan commented:
Sorry these files keep changing @anatasiavela -- I will try to review as soon as you complete the rebase (feel free to just merge trunk into it)

jolshan reviewed (approved):
thanks!

jolshan commented:
Seems like there was a build falilure related to `rat` I restarted the build.

jolshan commented:
Tests look unrelated. Going to merge.

anatasiavela commented:
@mimaison I've marked the JIRA as resolved

mimaison commented:
Thanks!
